File: program_stack/algorithms/searching/linear_search/cpp/linear_search.cpp
#include <algorithm>
#include <cstdlib>
#include <iostream>
#include <vector>

// Return index of first x in a, or -1 if absent.
int search(const std::vector<int>& a, int x) {
    for (std::size_t i = 0; i < a.size(); ++i) {
        if (a[i] == x) {
            return static_cast<int>(i);
        }
    }
    return -1;
}

static void expect(const std::vector<int>& a, int x, int want) {
    int got = search(a, x);
    if (got != want) {
        std::cerr << "search got " << got << " want " << want << "\n";
        std::exit(1);
    }
}

int main() {
    expect({}, 1, -1);
    expect({42}, 42, 0);
    expect({42}, 7, -1);
    expect({3, 1, 4, 1, 5}, 1, 1);
    expect({3, 1, 4, 1, 5}, 5, 4);
    expect({3, 1, 4, 1, 5}, 9, -1);
    std::cout << "linear_search: ok\n";
    return 0;
}

File: program_stack/algorithms/searching/linear_search/java/LinearSearch.java
public class LinearSearch {
    /** Return index of first x in a, or -1 if absent. */
    public static int search(int[] a, int x) {
        if (a == null) {
            return -1;
        }
        for (int i = 0; i < a.length; i++) {
            if (a[i] == x) {
                return i;
            }
        }
        return -1;
    }

    private static void expect(int[] a, int x, int want) {
        int got = search(a, x);
        if (got != want) {
            System.err.println("search got " + got + " want " + want);
            System.exit(1);
        }
    }

    public static void main(String[] args) {
        expect(new int[] {}, 1, -1);
        expect(new int[] {42}, 42, 0);
        expect(new int[] {42}, 7, -1);
        expect(new int[] {3, 1, 4, 1, 5}, 1, 1);
        expect(new int[] {3, 1, 4, 1, 5}, 5, 4);
        expect(new int[] {3, 1, 4, 1, 5}, 9, -1);
        System.out.println("linear_search: ok");
    }
}
